components/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java

```diff
@@ -1494,4 +1411,454 @@ private static Validator resourceToValidatorObject(
+                    urls = validatorType.equals(OCSP_VALIDATOR) ? getAIALocations(certificate) :
+                            getCRLUrls(certificate);
+                }
+            }
+        }
+        return urls;
+    }
+
+    private static void updateIssuerDNMap(IssuerDNMap issuerDNMap, String issuerDN, String serialNumber,
+                                          CertObject certObject)
+            throws CertificateValidationException {
+
+        List<CertObject> certList = issuerDNMap.getIssuerCertMap().computeIfAbsent(issuerDN, k -> new ArrayList<>());
+        boolean serialExists = certList.stream()
+                .anyMatch(cert -> getNormalizedName(cert.getSerialNumber()).equals(serialNumber));
+
+        if (serialExists) {
+            throw new CertificateValidationException("Certificate with serial number " + serialNumber +
+                    " already exists.");
+        }
+        certList.add(certObject);
+    }
+
+    private static CertObject saveUpdatedResource
+            (org.wso2.carbon.identity.configuration.mgt.core.model.Resource resource, IssuerDNMap issuerDNMap)
+            throws IOException, ConfigurationManagementException {
+
+        String serializedContent = ModelSerializer.serializeIssuerDNMap(issuerDNMap);
+        InputStream inputStreamUpdated = new ByteArrayInputStream(serializedContent.getBytes(StandardCharsets.UTF_8));
+
+        ResourceFile resourceFileUpdated = new ResourceFile();
+        resourceFileUpdated.setName(X509_CA_CERT_FILE);
+        resourceFileUpdated.setInputStream(inputStreamUpdated);
+        resource.getFiles().set(0, resourceFileUpdated);
+
+        CertValidationDataHolder.getInstance().getConfigurationManager()
+                .replaceResource(X509_CA_CERT_RESOURCE_TYPE, resource);
+
+        return new CertObject();
+    }
+
+    /**
+     * Get certificate from the configuration store by certificate id.
+     *
+     * @param tenantDomain Tenant Domain
+     * @param certificateId Certificate Id
+     * @return CACertificate
+     * @throws CertificateValidationException Error when getting certificate
+     */
+    public static CACertificate getCertificateFromConfigurationStoreByCertificateId(String tenantDomain,
+                                                                                    String certificateId)
+            throws CertificateValidationException {
+
+        try {
+            org.wso2.carbon.identity.configuration.mgt.core.model.Resource resource =
+                    CertValidationDataHolder.getInstance().getConfigurationManager().getResourceByTenantId
+                            (IdentityTenantUtil.getTenantId(tenantDomain), X509_CA_CERT_RESOURCE_TYPE, CERTS);
+            List<ResourceFile> resourceFiles = resource.getFiles();
+            if (resourceFiles == null || resourceFiles.isEmpty()) {
+                log.warn("Resource files are empty for certificates in tenant: " + resource.getTenantDomain());
```

Comment:
do we need the `WARN` log level

Reply:
if someone tried to get the cert list even though no cert is configured this might be a useful case.
But at the same time we can return empty list.
So will change them to debug logs